// bodhi-driver-superagent/lib/credentials.js
(function(exports){

    var BasicCredential = exports.BasicCredential = function(username, password){
        var credential = this;

        Object.defineProperty(credential, 'userAgent', {
            get: function() { return 'bodhi-user'},
            enumerable: true,
            configurable: false
        });

        credential.auth = function(request){
            request.auth(username, password);
            request.set('User-Agent', 'nodejs/' + process.version);
            return request;
        };
        credential.toHeaders = function toHeader(){
            return {
                auth: {
                    user: username,
                    pass: password,
                    sendImmediately: true
                },
                headers:{
                    'User-Agent': 'nodejs/' + process.version
                }
            };
        };
        credential.toString = function(){
            return username;
        };
    };

    var BearerToken = exports.BearerToken = function(token){
        var credential = this;
        var _agent     = 'bodhi-agent';

        Object.defineProperty(credential, 'userAgent', {
            get: function() { return _agent; },
            enumerable: true,
            configurable: false
        });

        credential.auth = function(request){
            request.set('Authorization', 'Bearer ' + token);
            return request;
        };
        credential.toHeaders = function toHeader(){
            return {
                auth: {
                    bearer: token,
                    sendImmediately: true
                },
                headers:{
                    'User-Agent': _agent
                }
            }
        };
        credential.toString = function(){
            return _agent;
        };
    };

    var CookieAuth = exports.CookieAuth = function(cookie){
        var credential = this;
        var _agent     = 'bodhi-agent';

        Object.defineProperty(credential, 'userAgent', {
            get: function() { return _agent; },
            enumerable: true,
            configurable: false
        });

        credential.auth = function(request){
            request.set('Cookie', cookie);
            return request;
        };

        credential.toHeaders = function toHeader(){
            return {
                auth: {
                    cookie: cookie,
                    sendImmediately: true
                },
                headers:{
                    'User-Agent': _agent
                }
            }
        };
        credential.toString = function(){
            return _agent;
        };
    };

})(module.exports)
