// bodhi-driver-superagent/lib/rest-patch.js
(function(exports){

    function ensure(path){

        path = path || '';

        return (0 !== path.indexOf('/')) ? "/" + path : path;
    }


    var Patch = exports.Patch = function Patch(basis){
        this.changes = [];
        this.basis   = basis || null;
    };

    Patch.prototype = {

        //add a new property entirely
        add : function(path, value){
            this.changes.push({op: 'add', path: ensure(path), value: value});
            return this;
        },

        //replace an existing property
        replace: function(path, value){
            this.changes.push({op: 'replace', path: ensure(path), value: value});
            return this;
        },

        //remove an existing property
        remove : function(path){
            this.changes.push({op: 'remove', path: ensure(path)});
            return this;
        },

        copy: function(path, from){
            this.changes.push({op: 'copy', path: ensure(path), from: ensure(from)});
            return this;
        },

        move: function(path, from){
            this.changes.push({op: 'move', path: ensure(path), from: ensure(from)});
            return this;
        },

        toJson : function(){
            return this.changes;
        }

    };


})(module.exports);
